docs/extensibility/custom-resources.md

```diff
@@ -0,0 +1,455 @@
+          "protocol": "tcp",
+          "transport": "tcp",
+          "targetPort": 1025
+        }
+      }
+    },
+    "newsletterservice": {
+      "type": "project.v0",
+      "path": "../MailDevResource.NewsletterService/MailDevResource.NewsletterService.csproj",
+      "env": {
+        "OTEL_DOTNET_EXPERIMENTAL_OTLP_EMIT_EXCEPTION_LOG_ATTRIBUTES": "true",
+        "OTEL_DOTNET_EXPERIMENTAL_OTLP_EMIT_EVENT_LOG_ATTRIBUTES": "true",
+        "OTEL_DOTNET_EXPERIMENTAL_OTLP_RETRY": "in_memory",
+        "ASPNETCORE_FORWARDEDHEADERS_ENABLED": "true",
+        "ConnectionStrings__maildev": "{maildev.connectionString}"
+      },
+      "bindings": {
+        "http": {
+          "scheme": "http",
+          "protocol": "tcp",
+          "transport": "http"
+        },
+        "https": {
+          "scheme": "https",
+          "protocol": "tcp",
+          "transport": "http"
+        }
+      }
+    }
+  }
+}
+```
+
+Because `MailDevResource` implements <xref:Aspire.Hosting.ApplicationModel.IResourceWithConnectionString> the manifest publishing logic in .NET Aspire knows that even though `MailDevResource` is a container resource, it also needs a `connectionString` field. The `connectionString` field references other parts of the `maildev` resource in the manifest to produce the final string:
+
+```json
+{
+    // ... other content omitted.
+    "connectionString": "smtp://{maildev.bindings.smtp.host}:{maildev.bindings.smtp.port}"
+}
+```
+
+.NET Aspire knows how to form this string because it looks at <xref:Aspire.Hosting.ApplicationModel.IResourceWithConnectionString.ConnectionStringExpression> and builds up the final string via the <xref:Aspire.Hosting.ApplicationModel.IManifestExpressionProvider> interface (in much the same way as the <xref:Aspire.Hosting.AppicationModel.IValueProvider> interface is used).
+
+The `MailDevResource` automatically gets included in the manifest because it is derived from <xref:Aspire.Hosting.ApplicationModel.ContainerResource>. Resource authors can choose to suppress outputting content to the manifest by overriding the default behavior using the following code (inside the `AddMailDev(...)` extension)
+
+```csharp
+public static IResourceBuilder<MailDevResource> AddMailDev(this IDistributedApplicationBuilder builder, string name, int? httpPort = null, int? smtpPort = null)
+{
+    var resource = new MailDevResource(name);
+    return builder.AddResource(resource)
+                    .WithImage(MailDevContainerImageTags.Image)
+                    .WithImageRegistry(MailDevContainerImageTags.Registry)
+                    .WithImageTag(MailDevContainerImageTags.Tag)
+                    .WithHttpEndpoint(targetPort: 1080, port: httpPort, name: MailDevResource.HttpEndpointName)
+                    .WithEndpoint(targetPort: 1025, port: smtpPort, name: MailDevResource.SmtpEndpointName)
+                    .WithAnnotation(ManifestPublishingCallbackAnnotation.Ignore); // This line added!
```

Comment:
Why this an not ExcludeFromManifest?

Reply:
Would you believe that I keep forgetting it exists. Fixed.